### PR TITLE
pygame.transform.blur

### PR DESCRIPTION
**Pull request by yunline**
# Add `blur` to transform module.
```python
def box_blur(
    surface: Surface,
    radius: int,
    repeat_edge_pixels: Optional[bool] = True,
    dest_surface: Optional[Surface] = None
) -> Surface: ...

def gaussian_blur( ... ) -> Surface: ...
def bokeh_blur( ... ) -> Surface: ...
```
Inspired by #3659.

image
☑ Box Blur
☑ Gaussian Blur
☐ Bokeh Blur

### CODE REVIEW

**Comment by novialriptide** [comment]
Looks like you need to run `python setup.py format` to pass the linting test.

File: src_c/transform.c [hunk 1]
```diff
@@ -2966,6 +2966,152 @@ surf_average_color(PyObject *self, PyObject *args, PyObject *kwargs)
     return Py_BuildValue("(bbbb)", r, g, b, a);
 }
 
+static void
+box_blur(SDL_Surface *src, SDL_Surface *dst, int radius)
```
Comment (yunline) [inline src_c/transform.c:2970]: It takes about 100ms to process a image using this algorithm.

Image: 2048x2048 RGBA
CPU: AMD Ryzen7 6800H

File: src_c/transform.c [hunk 2]
```diff
@@ -2966,6 +2966,152 @@ surf_average_color(PyObject *self, PyObject *args, PyObject *kwargs)
...
+    new_surf = blur(src_surf_obj, dst_surf_obj, radius);
+    if (!new_surf) {
+        return NULL;
+    }
+
+    if (dst_surf_obj) {
+        Py_INCREF(dst_surf_obj);
+        return (PyObject *)dst_surf_obj;
+    }
+    else {
```
Comment (novialriptide) [inline src_c/transform.c:3110]: Code quality nitpick: Could you remove the else statement and put the return new surface line at the end of the function?

**Review by novialriptide — CHANGES_REQUESTED** [review]

File: src_c/transform.c [hunk 3]
```diff
@@ -2966,6 +2966,152 @@ surf_average_color(PyObject *self, PyObject *args, PyObject *kwargs)
...
+}
+
+static SDL_Surface *
+blur(pgSurfaceObject *srcobj, pgSurfaceObject *dstobj, int radius)
+{
+    SDL_Surface *src = NULL;
+    SDL_Surface *retsurf = NULL;
+
+    if (radius < 0) {
+        return RAISE(PyExc_ValueError, "Radius less than zero.");
```
Comment (novialriptide) [inline src_c/transform.c:3045]: phrasing could be better like "the radius should not be less than zero"

**Review by novialriptide — CHANGES_REQUESTED** [review]

**Comment by novialriptide** [comment]
Blurring Better Call Saul is giving me weird results here... I'm not sure if this is normal

### Result
image

### Original Image
maxresdefault

**Comment by novialriptide** [comment]
You can never go wrong with Saul's good looks.

**Comment by yunline** [comment]
> It takes about 220ms to process an image using this algorithm.

I optimized the algorithm and now it takes 75ms to process one image.

Image: 2048x2048 RGBA
CPU: AMD Ryzen7 6800H

**Comment by itzpr3d4t0r** [comment]
I think that a blur function should be named depending on the blur algorithm it uses, so gaussian, bokeh etc.

**Comment by yunline** [comment]
New parameter `repeat_edge_pixels`
image

**Comment by yunline** [comment]
Gaussian blur √
image

**Pull request closed without merging**